### PR TITLE
Support UUID type in msgpack

### LINKED ISSUE #90: Is there support of built-in UUID type?
How can I insert uuid field from Golang?

### PR DESCRIPTION
**Pull request by DifferentialOrange**
This patch provides UUID support for all space operations and as function return result. UUID type was introduced in Tarantool 2.4.1. See more in [commit messages](https://github.com/tarantool/tarantool/commit/d68fc29246714eee505bc9bbcd84a02de17972c5).

To use UUID with google/uuid in msgpack, import tarantool/go-tarantool/uuid submodule.
```go
package main

import (
	"log"
	"time"

	"github.com/tarantool/go-tarantool"
	_ "github.com/tarantool/go-tarantool/uuid"
	"github.com/google/uuid"
)

func main() {
	server := "127.0.0.1:3013"
	opts := tarantool.Opts{
		Timeout:       500 * time.Millisecond,
		Reconnect:     1 * time.Second,
		MaxReconnects: 3,
		User:          "test",
		Pass:          "test",
	}
	client, err := tarantool.Connect(server, opts)
	if err != nil {
		log.Fatalf("Failed to connect: %s", err.Error())
	}

	spaceNo := uint32(524)
	indexNo := uint32(0)

	id, uuidErr := uuid.Parse("c8f0fa1f-da29-438c-a040-393f1126ad39")
	if uuidErr != nil {
		t.Errorf("Failed to prepare uuid: %s", uuidErr)
	}

	resp, err := conn.Replace(space, index, 0, 1, IterEq, []interface{}{ id })

	log.Println("UUID tuple replace")
	log.Println("Error", err)
	log.Println("Code", resp.Code)
	log.Println("Data", resp.Data)
}
```

Closes #90

### CODE REVIEW

**Comment by funny-falcon** [comment]
Could you place it into subfolder (subpackage)? I don't like it in main package.

**Comment by DifferentialOrange** [comment]
> Could you place it into subfolder (subpackage)? I don't like it in main package.

What's the motivation? UUID is built-in in Tarantool.

**Comment by funny-falcon** [comment]
> What's the motivation? UUID is built-in in Tarantool.

Yep. But it doesn't built-in in Golang.

Why github.com/google/uuid is better than any other?

- github.com/hashicorp/go-uuid
- github.com/satori/go.uuid
- github.com/gofrs/uuid
- go.mongodb.org/mongo-driver/x/mongo/driver/uuid

Looks like satori's one is more popular than google's one.

**Comment by DifferentialOrange** [comment]
> > What's the motivation? UUID is built-in in Tarantool.
> 
> Yep. But it doesn't built-in in Golang.
> 
> Why github.com/google/uuid is better than any other?
> 
>     * github.com/hashicorp/go-uuid
> 
>     * github.com/satori/go.uuid
> 
>     * github.com/gofrs/uuid
> 
>     * go.mongodb.org/mongo-driver/x/mongo/driver/uuid
> 
> 
> Looks like satori's one is more popular than google's one.

Well, I don't think it *is better* than any other package, but AFAIK it doesn't *worse* than any of them since they all implement similar basic functionality. Regarding popularity, satori/go.uuid is a bit more popular than google/uuid in terms of GitHub stars (4.4k vs 3.2k), but google/uuid leaves behind satori/go.uuid in case of `Used by` statistics: 59.1k GitHub repos vs 15.4k with satori/go.uuid. I think the latter is more important. To be honest, I have chosen google/uuid primarly because it is the package that was used by clients that requested UUID support for Tarantool in issue comments, so if you know any reasons why it's better not to use google/uuid or to use any other package, I would be glad to listen since I do not carried out a thorough comparison.

Yeah, if someone is willing to use a Tarantool connector with application that already uses another uuid package, it can introduce a conflict of namespaces. I'll try to rework it to avoid it.

**Comment by funny-falcon** [comment]
Yep, I'm not against google/uuid. And I'm not promoting any particular uuid package. I'm against including any particular uuid in main package-folder.

Golang provocates using subpackages (subfolders) for anything optional. And I think it is right way. And dependency on particular uuid package is optional.

**Comment by DifferentialOrange** [comment]
Moved to submodule in new commit

**Comment by ochaton** [comment]
I've noticed that without importing  `github.com/tarantool/go-tarantool/uuid` connector returns msgpack exception about undefined msg ext=2. This was made intentionally?

**Comment by LeonidVas** [comment]
Hi! Thank you for the patch. Generally ok.
See several comments bellow:
- I (and not only) have the habit of watching one patchset commit after another during a review and I see no reason to make changes in the first commit of a patchset and then change them in the next. If a different behavior is adopted in this project, then ok.
- what for we need a second `config.lua` in uuid that partially duplicates the `config.lua` from the root of the project?

**Comment by DifferentialOrange** [comment]
> I've noticed that without importing `github.com/tarantool/go-tarantool/uuid` connector returns msgpack exception about undefined msg ext=2. This was made intentionally?

Well, since uuid was extracted to submodule after review, you can say that it is intentional, but discussable

File: uuid/uuid_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	"gopkg.in/vmihailenco/msgpack.v2"
+	"github.com/google/uuid"
+)
+
+var server = "127.0.0.1:3013"
+var opts = Opts{
+	Timeout: 500 * time.Millisecond,
+	User:    "test",
+	Pass:    "test",
+	//Concurrency: 32,
```
Comment (LeonidVas) [inline uuid/uuid_test.go:19]: Remove the commented code or add a comment why it shouldn't be removed.

**Review by LeonidVas — APPROVED** [review]
LGTM, but I am not very familiar with the project and the rules for formatting the code and commits in it, so I could not take into account some points

File: uuid/uuid_test.go (hunk 1 above)
Comment (DifferentialOrange) [reply uuid/uuid_test.go]: Fixed

File: tarantool_test.go [hunk 2]
```diff
@@ -1005,3 +1006,119 @@ func TestComplexStructs(t *testing.T) {
...
+	resp, errSel := conn.Select(uuidSpace, uuidIndex, 0, 1, IterEq, []interface{}{ id })
+	if errSel != nil {
+		t.Errorf("UUID select failed: %s", errSel.Error())
+	}
+	if resp == nil {
+		t.Errorf("Response is nil after Select")
+	}
+	if len(resp.Data) != 1 {
+		t.Errorf("Response Data len != 1")
+	}
```
Comment (Totktonada) [inline tarantool_test.go]: We don't check data in the response of `Select()` except for length of the tuple. Is it intentional?

File: msgpack_ext.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	var bytesCount int = 16;
+	bytes := make([]byte, bytesCount)
+
+	n, err := d.Buffered().Read(bytes)
+	if err != nil {
+		return fmt.Errorf("msgpack: can't read bytes on uuid decode: %w", err)
+	}
+	if n < bytesCount {
+		return fmt.Errorf("msgpack: invalid uuid bytes count %d instead of %d", n, bytesCount)
+	}
```
Comment (Totktonada) [inline msgpack_ext.go]: Is it end of the stream or end of mp_ext bound?

**Review by Totktonada — APPROVED** [review]
Do you plan to squash commits (it seems worthful)? Shouldn't we drop github.com/google/uuid from go.mod in the second commit (and move it to go.mod inside `uuid/`)? Or it does not work this way?

I have a couple of questions, but generally can't add a value here. LGTM, but my review is unsufficient.

**Comment by DifferentialOrange** [comment]
> Do you plan to squash commits (it seems worthful)?

Yes, I changed in in a last force-push.

> Shouldn't we drop github.com/google/uuid from go.mod in the second commit (and move it to go.mod inside `uuid/`)?

Thank you, it definitely should be like this. Fixed.

File: uuid/config.lua [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    listen = 3013,
+    wal_dir = 'xlog',
+    snap_dir = 'snap',
+}
+
+box.schema.user.create('test', { password = 'test' , if_not_exists = true })
+box.schema.user.grant('test', 'execute', 'universe', nil, { if_not_exists = true })
+
+local uuid = require('uuid')
+local msgpack = require('msgpack')
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply uuid/config.lua]: I based the code style here on other `config.lua` scripts, but I don't think it is really meaningful, so I fixed it.

File: tarantool_test.go (hunk 2 above)
Comment (DifferentialOrange) [reply tarantool_test.go]: Added checks for tuples data in TestSelect and TestReplace cases

File: msgpack_ext.go (hunk 3 above)
Comment (DifferentialOrange) [reply msgpack_ext.go]: I've tried to run some manual checks and it looks like the answer is "end of the stream"

Comment (Totktonada) [reply msgpack_ext.go]: So the message should be like 'unexpected end of msgpack stream after N uuid bytes', what do you think?

Comment (DifferentialOrange) [reply msgpack_ext.go]: Seems legit. Fixed

**Review by funny-falcon — APPROVED** [review]
I don't think go.mod changes a necessary. But nonetheless, approve.

**Pull request merged**